pkg/controller/networkpolicy/clusternetworkpolicy.go

```diff
@@ -484,21 +561,136 @@ func serviceAccountNameToPodSelector(saName string) *metav1.LabelSelector {
+//	  "web,test,":       [ns1, ns2]
+//	  "web,production,": [ns3, ns4]
+func groupNamespacesByLabelValue(affectedNSAndLabels map[string]map[string]string, labelKeys []string) map[string][]string {
+	nsGroupedByLabelVal := map[string][]string{}
+	for ns, nsLabels := range affectedNSAndLabels {
+		if groupKey := getLabelValues(nsLabels, labelKeys); groupKey != labelValueUndefined {
+			nsGroupedByLabelVal[groupKey] = append(nsGroupedByLabelVal[groupKey], ns)
+		}
+	}
+	return nsGroupedByLabelVal
+}
+
+func getLabelValues(labels map[string]string, labelKeys []string) string {
+	key := ""
+	for _, k := range labelKeys {
+		if v, ok := labels[k]; !ok {
+			return labelValueUndefined
+		} else {
+			key += v + labelValueSeparater
+		}
+	}
+	return key
+}
+
+// labelKeyValPairsToSelector creates a LabelSelector based on a list of label keys
+// and their expected values.
+func labelKeyValPairsToSelector(labelKeys []string, labelValues string) *metav1.LabelSelector {
+	labelValuesSep := strings.Split(labelValues, labelValueSeparater)
+	labelMatchCriteria := map[string]string{}
+	for i := range labelKeys {
+		labelMatchCriteria[labelKeys[i]] = labelValuesSep[i]
+	}
+	return &metav1.LabelSelector{
+		MatchLabels: labelMatchCriteria,
+	}
+}
+
+// toAntreaPeerForSameLabelNamespaces computes the appliedToGroups and addressGroups for each
+// group of Namespaces who have the same values for the sameLabels keys.
+func (n *NetworkPolicyController) toAntreaPeerForSameLabelNamespaces(peer crdv1beta1.NetworkPolicyPeer,
+	np metav1.Object, atgPerAffectedNS map[string]*antreatypes.AppliedToGroup,
+	labelValues string,
+	namespacesByLabelValues []string) (*controlplane.NetworkPolicyPeer, []*antreatypes.AppliedToGroup, []*antreatypes.AddressGroup, sets.Set[string]) {
+	labelKeys := peer.Namespaces.SameLabels
+	var labelIdentities []uint32
+	uniqueLabelIDs := map[uint32]struct{}{}
+	clusterSetScopeSelectorKeys := sets.New[string]()
+	// select Namespaces who, for specific label keys, have the same values as the appliedTo Namespaces.
+	nsSelForSameLabels := labelKeyValPairsToSelector(labelKeys, labelValues)
+	addressGroups := []*antreatypes.AddressGroup{n.createAddressGroup("", peer.PodSelector, nsSelForSameLabels, peer.ExternalEntitySelector, nil)}
+	if n.stretchNPEnabled && peer.Scope == crdv1beta1.ScopeClusterSet {
+		newClusterSetScopeSelector := antreatypes.NewGroupSelector("", peer.PodSelector, nsSelForSameLabels, peer.ExternalEntitySelector, nil)
+		clusterSetScopeSelectorKeys.Insert(newClusterSetScopeSelector.NormalizedName)
+		// In addition to getting the matched Label Identity IDs, AddSelector also registers the selector
+		// with the labelIdentityInterface.
+		matchedLabelIDs := n.labelIdentityInterface.AddSelector(newClusterSetScopeSelector, internalNetworkPolicyKeyFunc(np))
+		for _, id := range matchedLabelIDs {
+			uniqueLabelIDs[id] = struct{}{}
+		}
+	}
+	for id := range uniqueLabelIDs {
+		labelIdentities = append(labelIdentities, id)
+	}
```

Comment:
Doesn't n.labelIdentityInterface.AddSelector return unique list? I feel it should return unique list, then  `labelIdentities` can be initialized directly after line 676.

Reply:
It does not return unique list. The reason is in normal ACNP cases (namely `toAntreaPeerForCRD`), there are a list of ingress peers that needs to be processed, so that dedup happens after all potential label identities are retrieved. Deduping at `n.labelIdentityInterface.AddSelector` level is going to be extra overhead in this case. However it might be negligible and could potentially make the code cleaner.